client/executor/common/src/wasm_runtime.rs

```diff
@@ -19,6 +19,38 @@
 use crate::error::Error;
 use sp_wasm_interface::Value;
 
+/// Call site.
+///
+/// Contains variants on how to resolve wasm function that will be invoked.
+pub enum CallSite<'a> {
+	/// Call function exported with this name.
+	///
+	/// Located function should have (u32, u32) -> u64 signature.
+	Export(&'a str),
+	/// Call function by reference from table.
+	///
+	/// Located function should have (u32, u32) -> u64 signature.
+	Table(u32),
+	/// Call function by reference from table through a wrapper.
+	///
+	/// Invoked function (`dispatcher_ref`) function
+	/// should have (u32, u32, u32) -> u64 signature.
+	///
+	/// `func` will be passed to the invoked function as a first argument.
```

Comment:
I am a bit wondering about the naming of `TableWithWrapper`. It could be more generic.
I mean `func` is only the first argument, so it could be named first argument.
Similarly `dispatcher_ref` could be call func or anyfunc index.
And `TableWithWrapper` could be `TableWithIndex` or `TableWithU32Arg`.

I guess it depends if we want this to be used in different contexts. Not sure it is a good idea, just talking.

Reply:
Yeah, this can be [ab]used for other cases, but not sure we want to encourage this :)